service/worker/scanner/history/scavenger.go

```diff
@@ -317,6 +332,46 @@ func (s *Scavenger) getPaginationFn(
 	}
 }
 
+func (s *Scavenger) cleanUpWorkflowPassedRetention(
+	ctx context.Context,
+	mutableState *persistencepb.WorkflowMutableState,
+) error {
+	executionInfo := mutableState.GetExecutionInfo()
+	ns, err := s.registry.GetNamespaceByID(namespace.ID(executionInfo.GetNamespaceId()))
+	switch err.(type) {
+	case *serviceerror.NotFound,
+		*serviceerror.NamespaceNotFound:
+		return nil
+	case nil:
+		// continue to delete
+	default:
+		return err
+	}
+
+	retention := ns.Retention()
+	finalUpdateTime := executionInfo.GetLastUpdateTime()
+	ttl := time.Now().UTC().Sub(timestamp.TimeValue(finalUpdateTime))
+	if ttl > 0 && ttl > retention+s.historyDataMinAge() {
+		_, err = s.adminClient.DeleteWorkflowExecution(ctx, &adminservice.DeleteWorkflowExecutionRequest{
+			Namespace: ns.Name().String(),
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: executionInfo.GetWorkflowId(),
+				RunId:      mutableState.GetExecutionState().GetRunId(),
+			},
+		})
+		if err != nil {
+			// This is experimental. Ignoring the error so it will not block the history scavenger.
```

Comment:
Ignoring error will result in corrupted workflow where mutable state exists but not history. So I think we should not ignore the error.